async/zigcoro.zig

```diff
@@ -1,123 +1,203 @@
+        event: threading.ResetEventSingle = .{},
+        args: Signature.ArgsT,
         result: Signature.ReturnT = undefined,
 
         pub fn run(task_: *xev.ThreadPool.Task) void {
             const task: *Self = @alignCast(@fieldParentPtr("_task", task_));
-            task.result = @call(.auto, func, task.args.*);
-            task.notif.notify() catch @panic("Unable to notify");
+            task.result = @call(.auto, func, task.args);
+            task.event.set();
         }
     };
 
     var newtask: TaskT = .{
-        .notif = Notification.init() catch @panic("Notification.init failed"),
-        .args = &args,
+        .args = args,
     };
-    defer newtask.notif.deinit();
-
     AsyncThread.current.thread_pool.schedule(xev.ThreadPool.Batch.from(&newtask._task));
-    newtask.notif.wait() catch @panic("Unable to wait for notification");
-    return newtask.result;
-}
+    newtask.event.wait();
 
-pub fn tick() void {
-    AsyncThread.current.executor.exec.tick();
+    return newtask.result;
 }
 
 pub fn sleep(ms: u64) !void {
     try aio.sleep(AsyncThread.current.executor, ms);
 }
 
-pub const Notification = struct {
-    inner: aio.AsyncNotification,
+pub const threading = struct {
+    const Waiter = struct {
+        frame: libcoro.Frame,
+        thread: *const AsyncThread,
+        next: ?*Waiter = null,
+    };
+
+    const WaiterQueue = queue_mpsc.Intrusive(Waiter);
+
+    pub const ResetEventSingle = struct {
+        const State = union(enum) {
+            unset,
+            waiting: *Waiter,
+            set,
+
+            const unset_state: State = .unset;
+            const set_state: State = .set;
+        };
+
+        waiter: std.atomic.Value(*const State) = std.atomic.Value(*const State).init(&State.unset_state),
 
-    pub fn init() !Notification {
+        pub fn isSet(self: *ResetEventSingle) bool {
+            return self.waiter.load(&State.set_state, .monotonic) == &State.set_state;
+        }
+
+        pub fn reset(self: *ResetEventSingle) void {
+            self.waiter.store(&State.unset_state, .monotonic);
+        }
+
+        pub fn set(self: *ResetEventSingle) void {
+            switch (self.waiter.swap(&State.set_state, .monotonic).*) {
+                .waiting => |waiter| {
+                    waiter.thread.waiters_queue.push(waiter);
+                    waiter.thread.wake();
+                },
+                else => {},
+            }
+        }
+
+        pub fn wait(self: *ResetEventSingle) void {
+            var waiter: Waiter = .{
+                .frame = libcoro.xframe(),
+                .thread = AsyncThread.current,
+            };
+            var new_state: State = .{
+                .waiting = &waiter,
+            };
+            if (self.waiter.cmpxchgStrong(&State.unset_state, &new_state, .monotonic, .monotonic) == null) {
+                libcoro.xsuspend();
+            }
+        }
+    };
+};
+
+pub const FrameAllocator = struct {
```

Comment:
who is using this allocator ? I didn't see any async code using it directly

Reply:
yeah it's not really used, I kept it because it could prove useful